vector/v.ppa/main.c

```diff
@@ -0,0 +1,349 @@
+            pts[idx].x = points->x[0];
+            pts[idx].y = points->y[0];
+            pts[idx].id = idx;
+            double coords[2] = {points->x[0], points->y[0]};
+            kdtree_insert(kdtree, coords, idx, 0);
+            idx++;
+        }
+    }
+
+    G_message(_("Number of points: %d"), n);
+    G_message(_("Method: %s"), method);
+
+    if (strcmp(method, "k") == 0) {
+        calculate_k_function(kdtree, pts, n, output_file);
+    }
+    else if (strcmp(method, "l") == 0) {
+        calculate_l_function(pts, n, output_file);
+    }
+    else if (strcmp(method, "f") == 0) {
+        calculate_f_function(pts, n, output_file);
+    }
+    else if (strcmp(method, "g") == 0) {
+        calculate_g_functoin(kdtree, pts, n, output_file);
+    }
+    else {
+        G_fatal_error(_("Method not implemented yet"));
+    }
+
+    // Free memory and close the vector map
+    free(pts);
+    Vect_close(&Map);
+
+    return 0;
+}
+
+/**
+ * Calculates the k-function for a given set of points using a k-d tree.
+ *
+ * @param kdtree The k-d tree used for nearest neighbor searches.
+ * @param points An array of Point structures representing the points.
+ * @param n The number of points in the array.
+ * @param output_file The path to the output file where the results will be
+ * saved.
+ */
+void calculate_k_function(struct kdtree *kdtree, struct Point *points, int n,
+                          const char *output_file)
+{
+    FILE *fp = fopen(output_file, "w");
+    if (fp == NULL) {
+        G_fatal_error(_("Unable to open output file <%s>"), output_file);
+    }
+
+    double max_dist = 0.0;
+#pragma omp parallel for reduction(max : max_dist)
+    for (int i = 0; i < n; i++) {
+        for (int j = 0; j < n; j++) {
+            if (i != j) {
+                double dist = euclidean_distance(&points[i], &points[j]);
+                if (dist > max_dist) {
+                    max_dist = dist;
+                }
+            }
+        }
+    }
+
+    G_message(_("Max distance: %f"), max_dist);
+
+    fprintf(fp, "Distance,K-value\n");
+    for (double d = 0; d <= max_dist; d += max_dist / 100) {
+        double k_value = 0.0;
+
+#pragma omp parallel for reduction(+ : k_value)
+        for (int i = 0; i < n; i++) {
+            double coords[2] = {points[i].x, points[i].y};
+            int *puid = NULL;
+            double *pd = NULL;
+            int count =
+                kdtree_dnn(kdtree, coords, &puid, &pd, d, &points[i].id);
+            k_value += count - 1; // subtract 1 to exclude the point itself
+            free(puid);
+            free(pd);
+        }
+
+        k_value /= (n * (n - 1));
+        fprintf(fp, "%f,%f\n", d, k_value);
+    }
+
+    kdtree_destroy(kdtree);
+    fclose(fp);
+}
+
+double euclidean_distance(const struct Point *p1, const struct Point *p2)
+{
+    return sqrt((p1->x - p2->x) * (p1->x - p2->x) +
+                (p1->y - p2->y) * (p1->y - p2->y));
+}
```

Comment:
Calculating the square root is computationally costly. You can speed up the module by returning the squared distance, i.e.
```
     return (p1->x - p2->x) * (p1->x - p2->x) +
                 (p1->y - p2->y) * (p1->y - p2->y);
```
and only calculate the square root when really needed, e.g. after you got the maximum squared distance, calculate only once the square root for the found max dist.

Reply:
Nice catch, I've updated the code.